alpenhorn/io/base.py

```diff
@@ -0,0 +1,173 @@
+"""BaseIO classes.
+
+Provides the basic infrastructure for StorageNode and StorageGroup I/O.
+
+These are very low-level classes.  Any module implementing the I/O class for
+something even remotely resembling a POSIX filesystem may be better served
+by subclassing from DefaultIO instead of from here directly.
+"""
+from __future__ import annotations
+from typing import TYPE_CHECKING
+
+import json
+import logging
+
+if TYPE_CHECKING:
+    from ..queue import FairMultiFIFOQueue
+    from ..storage import StorageNode, StorageGroup
+
+log = logging.getLogger(__name__)
+
+
+class BaseNodeIO:
+    """Base class for StorageNode I/O modules in alpenhorn.
+
+    Parameters
+    ----------
+    node : StorageNode
+        the node
+    """
+
+    # SETUP
+
+    def __init__(self, node: StorageNode) -> None:
+        self.node = node
+
+        # Load config if present
+        if node.io_config is None:
+            self.config = dict()
+        else:
+            self.config = json.loads(node.io_config)
+
+    def set_queue(self, queue: FairMultiFIFOQueue) -> None:
+        """Set the queue used for asynchronous I/O tasks."""
+        self._queue = queue
```

Comment:
I was in the middle of reviewing this when you added this review. This was something I was thinking as well - at the very least, `_queue` should be set to _something_ in the constructor so that we don't end up with an odd AttributeError, or there should be a `queue` property which raises a more informative exception if `_queue` hasn't been set.

Reply:
There's only one queue for the daemon as a whole (It's the task scheduler).  A new one can't be created for each StorageNode.

I think this question might be partly due to not having enough context for the full I/O framework.

The queue is never re-assigned, there's only ever one queue, created in the alpenhornd start-up.  Would it be better to assign the queue instance to a global and access it that way?

This part of the system doesn't seem very nice, and I'm not opposed to changing it somehow.

We could merge this call into the initialiser, but it wouldn't save any calls.  Instead of doing:
```
# Set the queue (I/O instance is automatically created the first time `node.io` is accessed)
node.io.set_queue(queue)
# Run an integrity check
node.io.check(filecopy)
```
it would be:
```
# Create the I/O instance
node.create_io(queue)
# Run an integrity check
node.io.check(filecopy)
```
_Most_ methods in the I/O framework don't use the queue, and don't need `set_queue` to be called, but I think in practice I think it's rare for I/O methods to be called without that happening.